Refactoring of the bolsh-arc64-event-logger branch

pavelvkozlov opened the pull request:
This PR is a result of refactoring of the [bolsh-arc64-event-logger](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/tree/bolsh-arc64-event-logger) branch. Here I've collected most of commits from the bolsh-arc64-event-logger branch. 

**What this branch do:**
The main idea of this branch is providing most valuable changes from the [bolsh-arc64-event-logger](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/tree/bolsh-arc64-event-logger) branch to the ```arc64``` branch. Some of commits in this branch have been modified (removed changes of clock freq and base address). But main code hasn't changed. 
- This branch provides change of the core clock frequency to 20MHz.

**What this branch doesn't do:**
I've dropped commits:
- [ARC: Low level Event logger for ARCv3](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/b7990355ad00df678ba6ced0e8bf605faf83e308)
- [ARCv2: Add PMU SLC support](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/260eccf326ca5de3d12411b29feb6ec92bfdaed8)
- [ARCv3: Add cluster PCTs driver](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/49cda94869225b927daa5a4241d488c899e37d09)

As they deserve their own branches, where they can be developed and prepared for merge into the ```arc64```
Also dropped commits:
- [ARC: Added check whether there is a L2$ in the system for HWPTW setup.](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/24fcaa1ada9fcc73f833619d86343cd74b066b50) (linked with event loger)
- [Change default clock frequency for core clock and UART clock to 30MHz.](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/ef67401a9bb3ad47606748257ffa5747bac645a9)

Also:
- This branch doesn't change base address to 0 for the HS5x, it will be done in separate branch. 

I have some doubts about commits:
[ARCv3: increase mmap address space for HS6x](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/0a21bd59e5afb1567f9f2482bd25eab8c177664a)
[ARCv3: Add the ability to change the size of mapped kernel address space](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/f41411a885ffbeed9765f2b68ec217b77d4f3931)

I think some parameters of System Memory Map can be changed, but as current variant was widly used and tested, I think I should offer my changes in a separate branch after merge of these changes.

Compatible with current buildroot. Tested on QEMU and nSIM arc32/arc64.

xxkent committed:
ARC: Add page attributes to the memory attribute register

MMU memory attribute register has to be the same for all cores.

Core 0 tunes this register in 2 places. Second time it adds
MEMATTR_IDX_UNCACHED and MEMATTR_VOLATILE properties. But others
cores(!=0) in the SMP configuration setuped only MEMATTR_NORMAL to this
register. This fix lets to have the same value in this register for all
cores.

xxkent committed:
ARC: Fix a mistake for I-cacche and D-cache invalidation.

xxkent committed:
ARC: Restore invalidate mode after flush operation

Fix an issue with using register ARC_REG_DC_CTRL

xxkent committed:
ARCv3: Skip r59 register for ARC64 processors when save/restore context

The r59 register is reserved for ARC64. Illegal operand exception will be
raised if it referenced.

xxkent committed:
ARCv3: Add ability to change size of the mapped kernel address space

This update add the ability to change the kernel address space from
1Gb up to 2Gb. Tested for HS5x/HS6x with MMUv6_32/MMUv6_48 and page
size=4K.

xxkent committed:
ARC: Fix an issue when changing RTP1 register and don't invalidate TLB

Add local_flush_tlb_all() function after changing RTP1 register from
early_pg_dir to swapper_pg_dir - main Linux kernel page table.

After this, we need to provide mapping for peripheral (UART for
example) to swapper_pg_dir which is done in early_fixmap_shutdown().

Before this patch there was an issue because we used cached UART
mapping from early_pg_dir but actual RTP1 pointed to new table
swapper_pg_dir.

This issue was found in QEMU because it doesn't simulate TLB while
on nSIM and RTL all worked fine because they simulate/use TLB.

xxkent committed:
ARCv3: increase mmap address space for HS6x

This fix was made to let 429.mcf from SPECint to allocate 1677Mb of memory.

xxkent committed:
ARC: Proper kernel mapping

Use proper page flags for different memory sections,
for example X is used only for text section.

Previously we mapped kernel as an one big chunk.

Now kernel mapping is page-based, so it requires to have
memblock initialized to be able to allocate page entries
dynamically.

xxkent committed:
ARCv3:DBG: Add debug feature to change clock frequency

For debug purpose now we can change clock frequency using a global
variable glb_clock_freq in devtree.c
This can be useful when we have timing violations in a bitfiles and need to
reduce clock. In this case we don't need to regenerate old kernels, just
changing glb_clock_freq in MDB.

Steps to change:
1. Get clock variable address from vmlinux file
 readelf -s vmlinux | grep glb_clock_freq
 33790: 85ed7484     4 OBJECT  GLOBAL DEFAULT   16 glb_clock_freq
2. Load loader file in MDB
3. Substract 0x80000000 value from address 0x85ed7484-0x80000000 = 0x05ed7484
4. Change glb_clock_freq at address 0x05ed7484
evalq *(unsigned int *)0x05ed7484 = 20000000      (20MHz)
5. Run target

xxkent committed:
ARCv3: Setup cluster apertures

There was an issue P10019796-55966 where we first time made an attempt
tuning cluster apertures. Issue was fixed. But the function which setups
apertures is commented out in the kernel. And all works.
Now I'm enabling and start testing the kernel with arc_cluster_mumbojumbo().
Not sure that it is necessary, will live will see.

xxkent committed:
ARCv3:DBG: add global variable to change UART baud in MDB

This patch adds the ability to change the UART baud using global
variable glb_uart_baud. This is for debugging/developing purposes.
Use this to set UART baud before starting a kernel right after load.

xxkent committed:
ARCv3: Fix a problem with perf event names

This patch fixes 8 char limitation to perf event names. New limit
is 16 bytes for a name, сorresponds to the limitation in HW.

pavelvkozlov committed:
ARCv3:haps: Set core and UART clock frequency to 20MHz

All recent bitfiles are built for 20MHz core clock.

Signed-off-by: Pavel Kozlov <pavel.kozlov@synopsys.com>

xxkent commented on arch/arc/include/asm/processor.h:
```diff
@@ -152,6 +152,6 @@ extern unsigned long __get_wchan(struct task_struct *p);
  * space during mmap's.
  */
-#define TASK_UNMAPPED_BASE      (TASK_SIZE / 3)
+#define TASK_UNMAPPED_BASE      0x10000000
```
Why do we use exactly 0x10000000, 256Mb? Let's try to use more universal approach using TASK_SIZE define.

pavelvkozlov replied:
I agree, this change is unneeded, let's move vm for mmap back (TASK_SIZE / 3) is a good value.

pavelvkozlov commented:
Dropped commit:  [ARCv3: Add option to select chunk size of kernel mapping.](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/65b4c6b5bf16de765e0278cc552b8bb7f9c88917)
This commit prevents kernel boot and causes to CI fails in configuration with 64K pages. 
It is open question should we add this commit to the ```arc64``` branch.  This commit offers tradeof of some benefits from using large blocks (2Mb PMD) and increasing the kernel image size. It doesn't work in all configurations and perhaps extra option should be added to the menuconfig to enable this feature. 
It should be dropped or reworked and fixed in separate branch.

pavelvkozlov commented:
Fixed build of arc64 configuration with 16K pages after drop of commit: [ARCv3: Add option to select chunk size of kernel mapping.](https://github.com/foss-for-synopsys-dwc-arc-processors/linux/commit/65b4c6b5bf16de765e0278cc552b8bb7f9c88917)
